### DISCUSSION TITLE
making a good Makefile

### DISCUSSION THREAD

**Discussion by alecristia** [LAAC-LSCP/ChildProject#256] [Q&A] [answered]
While documenting [how I set up an analysis folder for solomon](https://github.com/LAAC-LSCP/ChildProject/discussions/251), I'm drawing inspiration from EL1000's Makefile, which looks like this:

```
all: aclew_metrics.csv lena_metrics.csv

aclew_metrics.csv:
	datalad run -i "EL1000/*/annotations/alice/converted" \
	 -i "EL1000/*/annotations/vcm/converted" \
	 -i "EL1000/*/annotations/vtc/converted" \
	 -o aclew_metrics.csv "python code/metrics.py aclew aclew_metrics.csv"

lena_metrics.csv:
	datalad run -i "EL1000/*/annotations/its/converted" -o lena_metrics.csv "python code/metrics.py lena lena_metrics.csv"

clean:
	rm -f metrics.csv

```

One of the lines is similar to one part of the README.md: `datalad run -i "EL1000/*/annotations/its/converted" -o lena_metrics.csv "python code/metrics.py lena lena_metrics.csv"`

I get the syntax of this command: I'm telling datalad to run the command "python code/metrics.py lena lena_metrics.csv" with "EL1000/*/annotations/its/converted" as input parameter and "lena_metrics.csv" as output parameter.

I don't quite follow the Makefile one, though. This part I get:

- if user says "all", then run both subroutines

But this part I don't:

- if user says clean, then rm metrics.csv -- but metrics.csv hasn't been generated, since the other two are generating aclew_metrics.csv lena_metrics.csv

What am I missing?

In addition, my own folder will have at least two steps, one in which the metrics are calculated and a second one in which human-aclew reliability is calculated. I'm hesitant about adding a third step, which fits the actual statistical model used in the paper. This is a Bayesian model that takes a little while to fit; also, in my experience, those packages evolve and aren't always backward-compatible. Both of these reasons lead me to adding that code here, to make sure that this part is reproducible. But it goes against the habit of including the code of analysis in the paper itself. Any advice on this point?

**Comment by lucasgautheron** [comment]
Hello,

You're right about the Makefile actually, my mistake! Just fixed it.

Regarding your second question: it seems to me that your statistical model belongs to your paper dataset in any case (e.g. a YODA dataset); solis_metrics could be a subdataset of this dataset.

Also, a better example of Makefile is this one: https://gin.g-node.org/LAAC-LSCP/managing-storing-sharing-paper/src/main/Makefile
(I think it is actually better to keep 'datalad run' commands out of the Makefile).

**Comment by lucasgautheron** [comment, marked as answer]
Ok, so, for now your Makefile should look like this I'd say:

```Makefile
all: solomon.csv

solomon.csv: input/solomon/annotations/vtc/converted/*.csv input/solomon/annotations/alice/converted/*.csv
	child-project metrics input/solomon solomon.csv aclew

input/solomon/annotations/vtc/converted/*.csv:
	datalad get input/solomon/annotations/vtc/converted

input/solomon/annotations/alice/converted/*.csv:
	datalad get input/solomon/annotations/alice/converted

clean:
	rm -f solomon.csv
	datalad drop input/solomon/annotations/vtc/converted
	datalad drop input/solomon/annotations/alice/converted
```
